inst/tests/tests.Rraw

```diff
@@ -21084,3 +21084,18 @@ test(2307, { capture.output(print(DT, class = TRUE, show.indices = TRUE)); TRUE
 dt = data.table(date=as.IDate(c(NA, "2014-12-05")))
 test(2308.01, fread("date\nNA\n2014-12-05", keepLeadingZeros=TRUE), dt)
 test(2308.02, fread("date\nNA\n2014-12-05", keepLeadingZeros=FALSE), dt)
+
+# Test that as.data.table.data.table preserves key when explicitly specified but not when omitted
+# data.frame to data.table with key
+DF = data.frame(t = c(3:1, 4:5), y = 1:5)
+test(2309.01, key(as.data.table(DF, key="t")), "t")
+# tibble to data.table with key
+DF <- data.frame(t = c(3:1, 4:5), y = 1:5)
+class(DF) <- c("tbl_df", "tbl", "data.frame")
+test(2309.02, key(as.data.table(DF, key="t")), "t")
+# data.table to data.table with key
+DT = data.table(t = c(3:1, 4:5), y = 1:5)
+test(2309.03, key(as.data.table(DT, key="t")), "t")
+# data.table to data.table without key
+DT = data.table(t = c(3:1, 4:5), key="t")
+test(2309.04, key(as.data.table(DT)), NULL)
```

Comment:
Actually we have three cases:
For data.table keyed with "b":
* `as.data.table(key="a")` should set "a" as key
* `as.data.table()` should leave "b" as key intact
* `as.data.table(key=NULL)` should remove the key
For a non-keyed data.table:
* `as.data.table(key="a")` should set "a" as key
* `as.data.table()` should leave the data.table unkeyed
* `as.data.table(key=NULL)` should leave the data.table unkeyed